### ISSUE TITLE
add dedupe approach

### ISSUE THREAD

**Issue by cole-brokamp** [geomarker-io/parcel#4]
read in parcel data and make full addresses for matching (replace with tidyr::unite so missing doesn't show up in pasted string)

```r
readr::read_csv("cagis_parcels.csv") |>
  dplyr::mutate(parcel_address = paste(property_addr_number, property_addr_street, property_addr_suffix)) |>
  dplyr::select(-property_addr_number, -property_addr_street, -property_addr_suffix) |>
  na.omit() |>
  readr::write_csv("b.csv")
```

relies on python packages:

```r
system2("pip", "install csvdedupe")
system2("pip", "install dedupe-variable-address")
```

- second one is requred for "Address" field type
- does this also make postal available generally??

System call starts the process, but has to be done on CSV files.

```r
system2("csvlink", "a.csv b.csv --field_names_1 address --field_names_2 parcel_address --field-definition Address --inner_join --output_file test.csv")
```

This call would return only the matched rows.  Left join with the "input" CSV file (a.csv, where b.csv are the addresses from CAGIS) to return `NA` for those that were not matched.

Could we output to stdout instead of having to rewrite csv file?

thoughts:

- make these matching methods modular
- skip one if docker and/or python library isn't available

**Comment by cole-brokamp** [comment]
stashing a failed attempt at using reticulate here:

```r
library(reticulate)

# if you don't want to worry about your python setup and want to let reticulate manage it for you:
## py_install("dedupe", pip = TRUE)

# also install these optional addons to use variable type Address or Name
## py_install("dedupe-variable-address", pip = TRUE)
## py_install("dedupe-variable-name", pip = TRUE)

# https://docs.dedupe.io/en/latest/how-it-works/How-it-works.html

## reticulate::py_module_available("dedupe")
## dedupe <- import("dedupe")

# transform tibble into data structure expected by dedupe
d_ref <-
  d_ref |>
  as.list() |>
  purrr::transpose() |>
  rlang::set_names(d_ref$parcel_id)

d <-
  d |>
  as.list() |>
  purrr::transpose() |>
  rlang::set_names(d$PAT_ENC_CSN_ID)

#' read more about variable definitions: https://docs.dedupe.io/en/latest/Variable-definition.html#variable-definitions
var_def <-
  list(
    list("field" = "address", "type" = "Address")
    ## list("field" = "city", "type" = "String"),
    ## list("field" = "zip", "type" = "String")
  )

# try using crf = TRUE for String, ShortString, Address, and Name fields
# come back to looking at using "has missing" and "comparator" arguments
# try the "type" = "Interaction"

# create matcher based on variable definition
gaz <- dedupe$Gazetteer(variable_definition = var_def)

gaz$prepare_training(d, d_ref)

finished <- FALSE

while (!finished) {
  pair_q <- gaz$uncertain_pairs()[[1]]
  cli::cli_rule()
  cli::cli_alert_info("{length(gaz$training_pairs$match)}/10 positive, {length(gaz$training_pairs$distinct)}/10 negative \n")
  pair_q |>
    dplyr::bind_rows() |>
    knitr::kable() |>
    print()
  resp <- tolower(readline("Are these two things the same? \n (y)es, (n)o, (u)nsure, (f)inished) "))
  if (! resp %in% c("yes", "no", "unsure", "finished", "y", "n", "u", "f")) {
    cli::cli_alert_warning("please answer {.code y}, {.code n}, {.code u}, or {.code f}.")
    resp <- tolower(readline("Are these two things the same? \n (y)es, (n)o, (u)nsure, (f)inished) "))
    if (! resp %in% c("yes", "no", "unsure", "finished", "y", "n", "u", "f")) {
      cli::cli_alert_danger("answer {.code y}, {.code n}, {.code u}, or {.code f}.")
      stop()
    }
  }
  if (resp %in% c("y", "yes")) {
    gaz$mark_pairs(list("match" = list(pair_q),
                        "distinct" = reticulate::py_eval("[]", convert = FALSE)))
  }
  if (resp %in% c("n", "no")) {
    gaz$mark_pairs(list("distinct" = list(pair_q),
                        "match" = reticulate::py_eval("[]", convert = FALSE)))
  }
  if (resp == "unsure") {
    cli::cli_alert_info("unsure? skipping this one...")
  }
  if (resp %in% c("f", "finished")) {
    finished <- TRUE
  }
}

## dedupe$console_label(gaz)
gaz$train()
gaz$index(d_ref)

gaz$write_training(py_eval("open('training.json', 'w')"))
gaz$write_settings(py_eval("open('learned_settings', 'wb')"))

results <- gaz$search(d, threshold = 0, n_matches = 2, generator = FALSE)
  

## train_one_pair <- function(deduper) {
##   pair_q <- deduper$uncertain_pairs()[[1]]
##   cli::cli_rule()
##   cli::cli_alert_info("{length(deduper$training_pairs$match)}/10 positive, {length(deduper$training_pairs$distinct)}/10 negative \n")
##   pair_q |>
##     dplyr::bind_rows() |>
##     knitr::kable() |>
##     print()
##   resp <- tolower(readline("Are these two things the same? \n (y)es, (n)o, (u)nsure, (f)inished) "))
##   stopifnot(resp %in% c("yes", "no", "unsure", "finished", "y", "n", "u", "f"))
##   if (resp %in% c("f", "finished")) {
##     deduper$train()
##     cli::cli_alert_success("Finished training.")
##     return(invisible(deduper))
##   }
##   if (resp %in% c("y", "yes")) {
##   deduper$mark_pairs(list("match" = list(pair_q),
##                       "distinct" = reticulate::py_eval("[]", convert = FALSE)))
##   }
##   if (resp %in% c("n", "no")) {
##   deduper$mark_pairs(list("distinct" = list(pair_q),
##                       "match" = reticulate::py_eval("[]", convert = FALSE)))
##   }
##   if (resp == "unsure") {
##   }
##   return(invisible(deduper))
##   }

## ## train_one_pair(deduper = gaz)
## for (i in 1:20) gaz <- train_one_pair(deduper = gaz)

#### if training with active labeling
# add data to matcher instance
# Dedupe will find the next pair of records it is least certain about and ask you to label them as matches or not. use ‘y’, ‘n’ and ‘u’ keys to flag duplicates press ‘f’ when you are finished
## cli::cli_rule()
## cli::cli_alert_info("Starting active labeling.")
## cli::cli_text("Dedupe will find the next pair of records it is least certain about and ask you to label them as matches or not. Use `y`, `n`, and `u` keys to answer. Use `f` when you are finished.")
## cli::cli_rule()

## latest <- NULL
## while (is.null(latest)) latest <- train_one_pair(deduper = gaz)





## gaz$train()




gaz$write_training(py_eval("open('training.json', 'w')"))
gaz$write_settings(py_eval("open('learned_settings', 'wb')"))

# should return indexes of d_ref that are potential matches with each index of d (along with score)
results <- gaz$search(d, threshold = 0, n_matches = 2, generator = FALSE)
```

**Comment by cole-brokamp** [comment]
merged in #5
